File: Java-Project/DataStructure/src/datastructure/queue/PriorityQueue.java
package datastructure.queue;

import java.util.Arrays;

// [1, 7, 8] + 2
// [1, 2 , 7, 8]


public class PriorityQueue {
	private int[] items;
	private int front;
	private int rear;
	private final double extendBy = 2.0;
	
	public PriorityQueue() {
		items = new int[5];
		front = 0;
		rear = 0;
	}

	private void extendArray() {
		int[] array = new int[(int)(rear*extendBy)];
		int j=0;
		for(int i = front; i < rear; i++) {
			array[j] = items[i];
			j++;
		}
		front = 0;
		rear = j;
		items = array;
	}
	
	private boolean isEmpty() { return front == rear; }
	private boolean isFull() { return items.length == rear; }
	public void enqueue(int value) {
		if(isFull()) extendArray();
		items[rear] = value;
		rear++;
	}
	public int peek() {
		if(isEmpty()) throw new Error();
		return items[front];
	}
	public int dequeue() {
		int peek = peek();
		front++;
		return peek;
	}
	
	@Override
	public String toString() { return Arrays.toString( Arrays.copyOfRange(items, front, rear) ); }
}

File: CSharp-Project/DataStructureLinear/Queue/PriorityQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Queue
{
    public class PriorityQueue
    {
        private int[] Items { get; set; }
        private int Front { get; set; }
        private int Back { get; set; }

        public PriorityQueue()
        {
            Items = new int[10];
            Front = 0;
            Back = 0;
        }

        private bool IsEmpty() { return Front == Back; }
        private bool IsFull() { return Items.Length == Back + 1; }   // +1 !!!
        public void Enqueue(int value)
        {
            if (IsFull()) throw new Exception();
            int i;
            for (i = Back - 1; i >= Front; i--) //הזזה 
            {
                if (Items[i] > value) { Items[i + 1] = Items[i]; }
                else { break; }
            }
            Items[i+1] = value;
            Back++;
        }
        public int peek()
        {
            if (IsEmpty()) throw new Exception();
            return Items[Front];
        }
        public int dequeue()
        {
            int peek = Peek();
            Front++;    //!
            return peek;
        }

        public override String ToString() { return String.Join(",", Items[Front..Back])};
        }
    }
}
